util/auto_tune_compressor.cc

```diff
@@ -144,138 +152,222 @@ CostAwareCompressor::CostAwareCompressor(const CompressionOptions& opts)
+    return cur_compressor_idx_;
+  } else if (cur_compressor_idx_ >= compressors_.size()) {
+    // If the current compression type and level are not available i.e.
+    // Compression is disabled We can switch from no compression if we can
+    // decrease io and increase cpu usage else we current no compression is the
+    // best we can do
+    if (decrease_io && increase_cpu) {
+      cur_compressor_idx_ = 0;
+      return cur_compressor_idx_;
+    }
+    return cur_compressor_idx_;
+  }
 
-Status CostAwareCompressor::CompressBlockAndRecord(
-    size_t choosen_compression_type, size_t compression_level_ptr,
-    Slice uncompressed_data, std::string* compressed_output,
-    CompressionType* out_compression_type, CostAwareWorkingArea* wa) {
-  assert(choosen_compression_type < allcompressors_.size());
-  assert(compression_level_ptr <
-         allcompressors_[choosen_compression_type].size());
-  assert(choosen_compression_type < wa->cost_predictors_.size());
-  assert(compression_level_ptr <
-         wa->cost_predictors_[choosen_compression_type].size());
+  TEST_SYNC_POINT_CALLBACK(
+      "AutoTuneCompressorWrapper::CompressBlock::GetPredictors",
+      &(wa->cost_predictors_));
+  // If we are not in the stable region, then we need to explore the other
+  // compression algorithm and level which is in the right quadrant where we
+  // want to move to based on whether we want to increase or decrease the cpu
+  // and io usage
+  auto cur_cpu_cost =
+      wa->cost_predictors_[cur_compressor_idx_]->CPUPredictor.Predict();
+  auto cur_io_cost =
+      wa->cost_predictors_[cur_compressor_idx_]->IOPredictor.Predict();
+  for (size_t choice = 0; choice < compressors_.size(); choice++) {
+    auto predicted_io_cost =
+        wa->cost_predictors_[choice]->IOPredictor.Predict();
+    auto predicted_cpu_cost =
+        wa->cost_predictors_[choice]->CPUPredictor.Predict();
+    bool flag = true;
+    if (increase_io) {
+      if (predicted_io_cost <= cur_io_cost) {
+        flag = false;
+      }
+    }
+    if (decrease_io) {
+      if (predicted_io_cost >= cur_io_cost) {
+        flag = false;
+      }
+    }
+    if (increase_cpu) {
+      if (predicted_cpu_cost <= cur_cpu_cost) {
+        flag = false;
+      }
+    }
+    if (decrease_cpu) {
+      if (predicted_cpu_cost >= cur_cpu_cost) {
+        flag = false;
+      }
+    }
+    if (flag) {
+      cur_compressor_idx_ = choice;
+      return cur_compressor_idx_;
+    }
+  }
+  // If we did not find any other compression type and level in our intended
+  // quadrant, we may choose not to compress if we aim to increase IO and
+  // decrease CPU usage. Otherwise, the current compression type and level
+  // is the best we can do.
+  if (increase_io && decrease_cpu) {
+    // index that is above the size of created compressors is treated as signal
+    // for no compression
+    cur_compressor_idx_ = std::numeric_limits<size_t>::max();
+  }
+  return cur_compressor_idx_;
+}
```

Comment:
The chunk of logic is imperative and verbose, hard to read. It can use some refactoring. The high-level idea is (1) Compare current resource util and target range to figure out what direction to move in IO and CPU (2) Find the next compressor that satisfies that direction. Right not the coding is not reflecting the high-level idea clearly.